_posts/2025-11-19-wasm-ai-agents.adoc

```diff
@@ -0,0 +1,158 @@
+
+An *AI integration layer* takes care of actually integrating the LLM into our system, using https://github.com/langchain4j/langchain4j[LangChain4j] as a Java integration framework, https://github.com/jlama-ai/jlama[JLama] (Java implementation of LLaMA) for model inference, and TinyLlama-1.1B as a lightweight model for efficient local processing. LangChain4j provides seamless integration with both local and cloud-based language models, with a modular architecture that makes it easy to switch between different model providers. JLama provides optimized inference and memory management specifically designed for the JVM environment (both JLama and Chicory run entirely within the JVM boundaries, ensuring everything is self-contained within the Java ecosystem). TinyLlama-1.1B-Chat-v1.0 is a compact model that works well for demo purposes and can run efficiently on development machines, showing that local inference does not always require massive resources.
+
+[.text-center]
+.JVM as a Polyglot AI Agent Runtime
+image::wasm.png[width=50%, align="center", alt="JVM as a Polyglot AI Agent Runtime"]
+
+== The Polyglot Advantage
+
+Each language brings its strengths to the AI agent ecosystem:
+
+*Rust*: High-performance systems programming with memory safety, compiled to WebAssembly for maximum efficiency
+
+*Go*: Efficient and with clean syntax, leveraging TinyGo for WASM compilation with WASI support
+
+*Python*: Flexible scripting via PyO3 compilation to WASM, maintaining Python's expressiveness
+
+*JavaScript*: Dynamic scripting with QuickJS integration, enabling runtime flexibility
+
+The beauty of this approach is that all these languages run within the same JVM process, sharing resources and memory efficiently while maintaining their individual characteristics. This means that whatever language you used to build your agent can be integrated into this system, possibly together with other agents written in different languages.
+
+== Performance and Resource Efficiency
+
+The JVM approach offers several performance advantages:
+
+*Memory Efficiency*: The JVM's garbage collector manages memory for all languages uniformly, eliminating the memory management complexity of multi-runtime approaches.
+
+*Resource Sharing*: All agents share the same JVM process, reducing memory overhead and enabling efficient resource utilization.
+
+*Enterprise Monitoring*: Built-in JVM monitoring tools provide visibility into agent performance, memory usage, and execution patterns.
+
+*Scalability*: The stateless design allows for horizontal scaling across multiple JVM instances, with load balancing and container orchestration.
+
+=== Try It Yourself
+
+Getting started is straightforward with our Docker-based approach:
```

Comment:
I need to see code to understand - right now, I don't really understand what's in WASM module, what's outside and how the different things are implemented.